daemon/api.go

```diff
@@ -585,21 +585,50 @@ func deleteOp(c *Command, r *http.Request) Response {
 	}
 }
 
+type licT struct {
+	Intro   string
+	License string
+	Agreed  bool
+}
+
+func (*licT) Error() string {
+	// licT isn't an error, strictly speaking, but can
+	// be converted into one for convenience
+	return "license agreement required"
+}
+
 type packageInstruction struct {
+	progress.NullProgress
 	Action   string `json:"action"`
 	LeaveOld bool   `json:"leave_old"`
+	License  *licT  `json:"license"`
 	pkg      string
-	prog     progress.Meter
 }
 
+func (inst *packageInstruction) Agreed(intro, license string) bool {
+	if inst.License == nil || !inst.License.Agreed || inst.License.Intro != intro || inst.License.License != license {
+		inst.License = &licT{Intro: intro, License: license, Agreed: false}
+		return false
+	}
+
+	return true
+}
+
+var snappyInstall = snappy.Install
+
 func (inst *packageInstruction) install() interface{} {
 	flags := snappy.DoInstallGC
 	if inst.LeaveOld {
 		flags = 0
 	}
-	_, err := snappy.Install(inst.pkg, flags, inst.prog)
-
+	_, err := snappyInstall(inst.pkg, flags, inst)
 	if err != nil {
+		// to be more rigiorous we could also check
+		// err, ok := err.(*snappy.ErrInstallFailed); ok && err.OrigErr == snappy.ErrLicenseNotAccepted
+		// but that seems a bit overkill
```

Comment:
Your first paragraph makes me think you want a `snappy.IsLicenseNotAccepted(err)`, that would do the check described in the comment but be closer to the implementation details of `Install`. Maybe this is what you meant with your first comment, where you called it `IsLicensePending`; the `Pending` bit confused me but if this is what you meant, that makes sense and easy to do.

The second paragraph seems to contradict itself: if I feel strongly about *having* the “heuristics” (by which I've come to understand you mean the checking of the error's exact return value as in the comment), then surely removing the comment is the last thing I'd want to do. But maybe here you mean that either we do the thing, or we don't, but having the comment is wrong both ways?

I'll just nuke the comment.